Move ledger data:* favicons to external files

diracdeltas opened the pull request:
This moves data:*  favicon URLs in the ledger appState to external image files during session store cleanup. Reduces the size of my session-store-1 from ~23 MB to 5 MB.

Fix https://github.com/brave/browser-laptop/issues/11582

Test plan:
1. Copy session-store-1 from a Brave profile that has had payments enabled for a while to your brave-development appData directory.
2. Start the browser and go to about:preferences#payments. The synopsis table should appear fine.
3. Close the browser.
4. In your brave-development appData directory, you should now see a ledger-favicons subdirectory which contains a lot of image files. You should also notice that session-store-1 is now much smaller.
5. Repeat step 2.

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Test Plan:


Reviewer Checklist:

Tests


- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

NejcZdovc commented:
@diracdeltas this is really cool. Maybe one suggestion. Now we have synopsis and about synopsis, but synopsis is parent of about synopsis. So what is in about synopsis is for sure in synopsis. If I understand this code correctly we are create all favicons twice. My suggestion would be to generate favicons for synopsis and then just assign this values to about synopsis as well. wdyt?

NejcZdovc commented:
In near future we would like to remove this duplicate entries, so that you would only have one synopsis.

bridiver commented on app/sessionStore.js:
```diff
@@ -447,6 +459,42 @@ module.exports.cleanSessionDataOnShutdown = () => {
+          )
+          fs.writeFile(faviconPath, parsed.data, 'base64', (err) => {
+            if (err) {
+              console.error(`Error writing file: ${faviconPath} ${err}`)
```
are we ok with potentially losing these?

diracdeltas replied:
iirc the ledger tries to get the favicon whenever the user visits the page, so it would be re-fetched on the next visit

NejcZdovc replied:
@diracdeltas that's not the case. We only get favicon if there is not one

diracdeltas replied:
i could add an img onerror in the synopsis table that sets the faviconURL to `'' / null` if the file does not exist on disk - would that cause the favicon to be re-fetched?

NejcZdovc replied:
yes if you set it to null it will work. More info here: https://github.com/brave/browser-laptop/blob/master/app/browser/api/ledger.js#L1146

bridiver commented on app/sessionStore.js:
```diff
@@ -428,6 +429,17 @@ module.exports.cleanAppData = (immutableData, isShutdown) => {
+  try {
+    const basePath = getStoragePath('ledger-favicons')
+    const fs = require('fs')
+    if (!fs.existsSync(basePath)) {
```
should use the async versions of these to avoid blocking the UI thread

diracdeltas replied:
it needs to be synchnrous because this directory must exist before `cleanFavicons` is called, and `cleanFavicons` returns immutableData synchronously. I could add a migration flag that indicates whether this directory was created, which would avoid having to do the synchronous mkdir in future runs.

bsclifton commented:
For review status, @bridiver @NejcZdovc do you approve / not approve?  @diracdeltas was any action needed on your part after the above?

diracdeltas commented:
@bsclifton there's some follow-up enhancements which i'll try to do early this week

bsclifton commented:
Moving to 0.20.x (as the next 0.19.x release is reserved for Chromium 63)

diracdeltas commented on app/sessionStore.js:
```diff
@@ -447,6 +459,42 @@ module.exports.cleanSessionDataOnShutdown = () => {
+              ? `favicon-${index}-${Date.now()}.${parsed.ext}`
+              : `favicon-${index}.${parsed.ext}`
+          )
+          fs.writeFile(faviconPath, parsed.data, 'base64', (err) => {
```
fs.writeFile was easier because it supported the base64 encoding parameter

diracdeltas commented:
@NejcZdovc i addressed all review comments. however, i thought of something which is that the `ledger-favicon/` directory should be deleted whenever the user chooses to delete payments data. do you remember if that's doable via the 'Clear browsing data...' menu item or something else in the UI?

NejcZdovc commented:
we have issue up for this option https://github.com/brave/browser-laptop/issues/8537, so I think that we should finally implement it :smiley: will take a look and probably add it

diracdeltas commented:
ok, if there's currently no way to clear ledger data from the session store in the UI, i don't think this PR needs to address that. ready for re-review.

diracdeltas commented:
master: 4e7995dd468f7e6987a0ceca3938e9d0bc6fdfbe
0.21.x: ca4bc9af0ace764bc4562e8b0f2c3b79f50cfc32
0.20.x: 0b45c879ad08d9dc70bb7eb041665d0c20a2576b

petemill commented:
Do we have any sample large profiles with this data in, so we can assess impact and clear this change from any perf regressions we're seeing? @diracdeltas @NejcZdovc
